Flaky test: Data race in TestNetworkSendErrors/network-down

https://github.com/tailscale/tailscale/actions/runs/28211350454/job/83573170202?pr=19642

```
=== RUN   TestNetworkSendErrors/network-down
==================
WARNING: DATA RACE
Write at 0x000002119c8e by goroutine 271:
  tailscale.com/tstest.Replace[go.shape.bool]()
      /home/runner/work/tailscale/tailscale/src/tstest/tstest.go:46 +0x131
  tailscale.com/wgengine/magicsock.TestNetworkSendErrors.func1()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock_test.go:3384 +0x58
  testing.tRunner()
      /home/runner/.cache/tsgo/c803676bcc7f2b195b167a53d49d728045cd9b36/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/c803676bcc7f2b195b167a53d49d728045cd9b36/src/testing/testing.go:2101 +0x38

Previous read at 0x000002119c8e by goroutine 205:
  tailscale.com/wgengine/magicsock.(*Conn).networkDown()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock.go:1492 +0x58
  tailscale.com/wgengine/magicsock.(*Conn).runDerpReader()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/derp.go:573 +0x18c5
  tailscale.com/wgengine/magicsock.(*Conn).derpWriteChanForRegion.gowrap2()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/derp.go:462 +0x84

Goroutine 271 (running) created at:
  testing.(*T).Run()
      /home/runner/.cache/tsgo/c803676bcc7f2b195b167a53d49d728045cd9b36/src/testing/testing.go:2101 +0xb12
  tailscale.com/wgengine/magicsock.TestNetworkSendErrors()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock_test.go:3370 +0x39
  testing.tRunner()
      /home/runner/.cache/tsgo/c803676bcc7f2b195b167a53d49d728045cd9b36/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/c803676bcc7f2b195b167a53d49d728045cd9b36/src/testing/testing.go:2101 +0x38

Goroutine 205 (finished) created at:
  tailscale.com/wgengine/magicsock.(*Conn).derpWriteChanForRegion()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/derp.go:462 +0x1892
  tailscale.com/wgengine/magicsock.(*Conn).goDerpConnect.gowrap1()
      /home/runner/work/tailscale/tailscale/src/wgengine/magicsock/derp.go:313 +0x44
==================
Warning: sock.go:4118: magicsock: [warning] failed to force-set UDP read buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
Warning: sock.go:4118: magicsock: [warning] failed to force-set UDP write buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
Warning: sock.go:4118: magicsock: [warning] failed to force-set UDP read buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
Warning: sock.go:4118: magicsock: [warning] failed to force-set UDP write buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
## WARNING: (non-fatal) nil health.Tracker (being strict in CI):
goroutine 290 [running]:
tailscale.com/health.(*Tracker).nil(0x0)
	/home/runner/work/tailscale/tailscale/src/health/health.go:347 +0x70
tailscale.com/health.(*Tracker).SetUDP4Unbound(0x0, 0x0)
	/home/runner/work/tailscale/tailscale/src/health/health.go:933 +0x2f
tailscale.com/wgengine/magicsock.(*Conn).bindSocket(0xc004a52008, 0xc004a520b0, {0x16ad898, 0x4}, 0x0)
	/home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock.go:3791 +0x13da
tailscale.com/wgengine/magicsock.(*Conn).rebind(0xc004a52008, 0x0)
	/home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock.go:3820 +0x177
tailscale.com/wgengine/magicsock.NewConn({0xc00017e480, 0xc000052990, 0x0, 0x0, 0x0, 0x0, {0x0, 0x0}, 0xc000171110, 0x0, ...})
	/home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock.go:709 +0xfd7
tailscale.com/wgengine/magicsock.newTestConnAndRegistry(0xc004a08488)
	/home/runner/work/tailscale/tailscale/src/wgengine/magicsock/magicsock_test.go:3354 +0x2ae
tailscale.com/wgengine/magicsock.TestNetw
    peermtu.go:95: [v1] magicsock: peermtu: peer MTU status is false
    magicsock.go:740: magicsock: disco key = d:c393a7c6f1c9738b
    magicsock.go:2721: magicsock: SetNetworkUp(false)
    testing.go:1712: race detected during execution of test
--- FAIL: TestNetworkSendErrors/network-down (0.01s)
```

--- a/wgengine/magicsock/magicsock.go
+++ b/wgengine/magicsock/magicsock.go
@@ -425,6 +425,13 @@ type Conn struct {
 	// homeDERPGauge is the usermetric gauge for the home DERP region ID.
 	// This can be nil when [Options.Metrics] are not enabled.
 	homeDERPGauge *usermetric.Gauge
+
+	// checkNetworkUpDuringTests controls whether [Conn.networkDown]
+	// will report the value of [Conn.networkUp] while running tests.
+	//
+	// This allows tests to pass when the user's machine is offline,
+	// but allows us to still test network-down behaviour when desired.
+	checkNetworkUpDuringTests bool
 }
 
 // SetDebugLoggingEnabled controls whether spammy debug logging is enabled.
@@ -1482,14 +1489,10 @@ func (c *Conn) LocalPort() uint16 {
 
 var errNetworkDown = errors.New("magicsock: network down")
 
-// This allows tests to pass when the user's machine is offline, but allows us
-// to still test network-down behaviour when desired.
-var checkNetworkDownDuringTests = false
-
 func (c *Conn) networkDown() bool {
 	// For tests, always assume the network is up unless we're explicitly
 	// testing this behaviour.
-	if envknob.AssumeNetworkUp() || (testenv.InTest() && !checkNetworkDownDuringTests) {
+	if envknob.AssumeNetworkUp() || (testenv.InTest() && !c.checkNetworkUpDuringTests) {
 		return false
 	}
 	return !c.networkUp.Load()

--- a/wgengine/magicsock/magicsock_test.go
+++ b/wgengine/magicsock/magicsock_test.go
@@ -3386,9 +3386,8 @@ func TestNetworkSendErrors(t *testing.T) {
 			t.Skipf("skipping on %s", runtime.GOOS)
 		}
 
-		tstest.Replace(t, &checkNetworkDownDuringTests, true)
-
 		conn, reg := newTestConnAndRegistry(t)
+		conn.checkNetworkUpDuringTests = true
 
 		buffs := [][]byte{{00, 00, 00, 00, 00, 00, 00, 00}}
 		ep := &lazyEndpoint{

